By Dataquest and Vik Paruchuri 

In [1]:
import pandas as pd

In [2]:
games = pd.read_csv("nba_games.csv", index_col=0)

In [3]:
games.head()

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
0  240   240  39   81  0.481   6   20  0.300  14   18  ...          22.8   
1  240   240  36  100  0.360   7   31  0.226  16   19  ...          50.0   
2  240   240  37   85  0.435   8   19  0.421  17   23  ...          20.0   
3  240   240  41   89  0.461   8   21  0.381  17   19  ...          28.6   
4  240   240  27   86  0.314   6   26  0.231  15   20  ...          16.8   

   usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  \
0          29.0           178           111       DAL         95         1   
1          32.6           152           111       ATL         98         0   
2          30.9           148           116       SAS        107         1   
3          30.9           138           118       MIN         99         0   
4          30.9           157            90       MEM         92         1   

   season        date    won  
0    2016  09/12/2015   True  
1    2016  09/12/2015  False  
2    2018  18/10/2017  False  
3    2018  18/10/2017   True  
4    2021  30/04/2021  False  

[5 rows x 150 columns]

In [4]:
games.shape

(17772, 150)

In [5]:
games["team"].value_counts()

GSW    655
BOS    653
TOR    635
MIL    628
MIA    617
CLE    610
HOU    610
LAC    607
PHI    605
UTA    605
DEN    603
POR    603
OKC    595
DAL    593
SAS    593
ATL    590
PHO    590
MEM    585
IND    583
LAL    581
BRK    580
WAS    580
NOP    571
ORL    565
CHI    558
MIN    558
CHO    556
DET    556
SAC    554
NYK    553
Name: team, dtype: int64

In [6]:
games[games["team"] == "GSW"].sort_values("date")

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
15567  240   240  31   79  0.392   7   35  0.200  29   33  ...          50.0   
16686  240   240  46   86  0.535  17   32  0.531  14   21  ...          17.6   
1633   240   240  47   96  0.490  21   48  0.438  11   14  ...          29.4   
1196   240   240  48   90  0.533  18   35  0.514  17   19  ...          27.4   
13810  240   240  43   96  0.448  20   46  0.435  18   18  ...          23.1   
...    ...   ...  ..  ...    ...  ..  ...    ...  ..  ...  ...           ...   
2358   240   240  46   84  0.548  17   30  0.567  25   35  ...          33.3   
12751  240   240  47   89  0.528  16   32  0.500  21   30  ...          31.6   
10092  240   240  43   84  0.512  11   23  0.478  17   27  ...          20.5   
8348   240   240  46   84  0.548  14   28  0.500  26   32  ...          50.0   
14560  265   265  44   96  0.458  12   27  0.444  13   18  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
15567          29.0           250           102       POR        123   
16686          29.2           163           127       UTA        116   
1633           30.8           231           131       CHO        111   
1196           28.7           139           139       CLE        112   
13810          34.6           230           138       SAS        120   
...             ...           ...           ...       ...        ...   
2358           36.1           218           131       NOP        120   
12751          26.8           149           132       NOP        121   
10092          34.3           157           124       HOU        110   
8348           44.5           137           134       PHO        109   
14560          25.9           175           120       SAS        117   

       home_opp  season        date    won  
15567         0    2021  01/01/2021  False  
16686         1    2022  01/01/2022   True  
1633          0    2017  01/02/2017   True  
1196          1    2020  01/02/2020   True  
13810         1    2022  01/02/2022   True  
...         ...     ...         ...    ...  
2358          1    2016  31/10/2015   True  
12751         0    2019  31/10/2018   True  
10092         1    2016  31/12/2015   True  
8348          1    2019  31/12/2018   True  
14560         1    2020  31/12/2019  False  

[655 rows x 150 columns]

In [7]:
games["season"].value_counts()

2022    2646
2016    2632
2018    2624
2019    2624
2017    2618
2021    2342
2020    2286
Name: season, dtype: int64

In [8]:
games.dtypes

mp             int64
mp.1           int64
fg             int64
fga            int64
fg%          float64
              ...   
total_opp      int64
home_opp       int64
season         int64
date          object
won             bool
Length: 150, dtype: object

In [9]:
del games["mp.1"]

In [10]:
del games["+/-"]

converting existing column to a date time then overiding the existing column with the date time.

In [11]:
games["date"] = pd.to_datetime(games["date"], format="%d/%m/%Y")

In [12]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  tov%_max_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...          22.8   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...          50.0   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...          20.0   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...          28.6   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...          16.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...          34.2   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...          25.0   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...          25.6   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...          27.7   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...          51.5   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              29.0           178           111       DAL         95   
1              32.6           152           111       ATL         98   
2              30.9           148           116       SAS        107   
3              30.9           138           118       MIN         99   
4              30.9           157            90       MEM         92   
...             ...           ...           ...       ...        ...   
17767          33.7           160           118       OKC         92   
17768          30.0           139           129       ORL        108   
17769          29.9           175           126       LAC        113   
17770          27.1           150           126       MIA        106   
17771          36.2           141           114       BOS        117   

       home_opp  season       date    won  
0             1    2016 2015-12-09   True  
1             0    2016 2015-12-09  False  
2             1    2018 2017-10-18  False  
3             0    2018 2017-10-18   True  
4             1    2021 2021-04-30  False  
...         ...     ...        ...    ...  
17767         0    2019 2018-10-19   True  
17768         1    2017 2016-12-14   True  
17769         0    2017 2016-12-14  False  
17770         1    2020 2020-09-19   True  
17771         0    2020 2020-09-19  False  

[17772 rows x 148 columns]

In [13]:
games.dtypes

mp                    int64
fg                    int64
fga                   int64
fg%                 float64
3p                    int64
                  ...      
total_opp             int64
home_opp              int64
season                int64
date         datetime64[ns]
won                    bool
Length: 148, dtype: object

Creating predictors for machine learning 

home says whether the team played at a home game or an away game

In [14]:
games ["venue_code"] = games["home"].astype("category").cat.codes

Each opponent now has its own code 

In [15]:
games ["opp_code"] = games["team_opp"].astype("category").cat.codes

In [16]:
games ["day_code"] = games["date"].dt.dayofweek

In [17]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  drtg_max_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...           111   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...           111   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...           116   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...           118   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...            90   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...           118   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...           129   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...           126   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...           126   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...           114   

       team_opp  total_opp  home_opp  season       date    won  venue_code  \
0           DAL         95         1    2016 2015-12-09   True           0   
1           ATL         98         0    2016 2015-12-09  False           1   
2           SAS        107         1    2018 2017-10-18  False           0   
3           MIN         99         0    2018 2017-10-18   True           1   
4           MEM         92         1    2021 2021-04-30  False           0   
...         ...        ...       ...     ...        ...    ...         ...   
17767       OKC         92         0    2019 2018-10-19   True           1   
17768       ORL        108         1    2017 2016-12-14   True           0   
17769       LAC        113         0    2017 2016-12-14  False           1   
17770       MIA        106         1    2020 2020-09-19   True           0   
17771       BOS        117         0    2020 2020-09-19  False           1   

       opp_code  day_code  
0             6         2  
1             0         2  
2            26         2  
3            17         2  
4            14         4  
...         ...       ...  
17767        20         4  
17768        21         2  
17769        12         2  
17770        15         5  
17771         1         5  

[17772 rows x 151 columns]

The following code sets up a target to predict

In [18]:
games["target"] = (games["won"] == True).astype("int")

In [19]:
games 

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
0      240  39   81  0.481   6   20  0.300  14   18  0.778  ...       DAL   
1      240  36  100  0.360   7   31  0.226  16   19  0.842  ...       ATL   
2      240  37   85  0.435   8   19  0.421  17   23  0.739  ...       SAS   
3      240  41   89  0.461   8   21  0.381  17   19  0.895  ...       MIN   
4      240  27   86  0.314   6   26  0.231  15   20  0.750  ...       MEM   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17767  240  35   81  0.432  11   26  0.423  27   36  0.750  ...       OKC   
17768  240  37   74  0.500  13   25  0.520  26   37  0.703  ...       ORL   
17769  240  42   89  0.472  14   33  0.424  10   20  0.500  ...       LAC   
17770  240  41   85  0.482   9   26  0.346  26   30  0.867  ...       MIA   
17771  240  33   85  0.388  12   44  0.273  28   34  0.824  ...       BOS   

       total_opp  home_opp  season       date    won  venue_code  opp_code  \
0             95         1    2016 2015-12-09   True           0         6   
1             98         0    2016 2015-12-09  False           1         0   
2            107         1    2018 2017-10-18  False           0        26   
3             99         0    2018 2017-10-18   True           1        17   
4             92         1    2021 2021-04-30  False           0        14   
...          ...       ...     ...        ...    ...         ...       ...   
17767         92         0    2019 2018-10-19   True           1        20   
17768        108         1    2017 2016-12-14   True           0        21   
17769        113         0    2017 2016-12-14  False           1        12   
17770        106         1    2020 2020-09-19   True           0        15   
17771        117         0    2020 2020-09-19  False           1         1   

       day_code  target  
0             2       1  
1             2       0  
2             2       0  
3             2       1  
4             4       0  
...         ...     ...  
17767         4       1  
17768         2       1  
17769         2       0  
17770         5       1  
17771         5       0  

[17772 rows x 152 columns]

training Random Forest Model - this type of classifier can pick up non linear data - for example how the team codes dont show which team is better its just their assigned codes 

In [20]:
from sklearn.ensemble import RandomForestClassifier 

n_estiomators the number of individual decision trees I want to train. Each tree in the series of trees having slightly different parameters, the higher n_estiomators number is the longer it will take to run but the potentially more accurate it will be. The higher  min_samples_split is the less likely we are to overfit but the lower the accuracy on training data will be. There are a lot of random parameters in a random forest if I set a random state when we run the model multiple times on the same data we will get the same results 

In [21]:
rf = RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)

split up training and test data this data is time series dependent to make sure that all the data in the test set comes after all the data in the training set because data from the past has to be used to predict data in the future 

In [22]:
train = games[games["date"] < '2022-01-01']

In [23]:
test = games[games["date"] > '2022-01-01']

In [24]:
predictors = ["venue_code","opp_code","day_code"]

Fitting the random forest model with trained predictors to try and predict the target the target is 0 if the team lost or 1 if the team won

In [25]:
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

passing in the test data and the predictors 

In [26]:
preds = rf.predict(test[predictors])

determining the accuracy of the model it looks at what percentage of the time was your model correct in the prediction it made with either a win or a loss and then combining them to see what percentage of the time the model was actually accurate.

In [27]:
from sklearn.metrics import accuracy_score  

pass in our actuals and the predictions

In [28]:
acc = accuracy_score(test["target"],preds)

In [29]:
acc

0.5215189873417722

In [30]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))

In [31]:
pd.crosstab(index= combined["actual"],columns=combined["prediction"])

prediction    0    1
actual              
0           401  389
1           367  423

401 times the model was correct, 
367 times the model was wrong 

In [32]:
from sklearn.metrics import precision_score 

when a win was predicted the team only won 52% of the time which is low 

In [33]:
precision_score(test["target"], preds)

0.520935960591133

improvement of the model 

Improving the precision with rolling averages- by splitting the data set up in teams 

This groups the data set into team statistics

In [34]:
grouped_games = games.groupby("team")

In [35]:
group = grouped_games.get_group("GSW")

In [36]:
group

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
18     240  46   97  0.474  19   42  0.452   9   15  0.600  ...       ORL   
27     240  42   80  0.525  15   38  0.395  17   24  0.708  ...       PHI   
77     240  43   91  0.473  12   21  0.571  11   13  0.846  ...       DET   
112    240  33   95  0.347  16   48  0.333  17   24  0.708  ...       MIL   
118    240  34   94  0.362  10   25  0.400  17   23  0.739  ...       DET   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17688  240  38   78  0.487  12   28  0.429  20   24  0.833  ...       POR   
17694  240  43   92  0.467  10   40  0.250  15   21  0.714  ...       ATL   
17698  240  51   93  0.548  15   35  0.429  15   18  0.833  ...       LAC   
17713  265  53  102  0.520  14   37  0.378  14   15  0.933  ...       HOU   
17722  240  37   89  0.416  14   40  0.350  24   29  0.828  ...       PHO   

       total_opp  home_opp  season       date    won  venue_code  opp_code  \
18           124         1    2021 2021-02-19  False           0        21   
27            96         0    2022 2021-11-24   True           1        22   
77            95         0    2016 2015-11-09   True           1         8   
112          118         1    2022 2022-01-13  False           0        16   
118          113         1    2016 2016-01-16  False           0         8   
...          ...       ...     ...        ...    ...         ...       ...   
17688        125         1    2018 2018-03-09  False           0        24   
17694        117         1    2021 2021-04-04  False           0         0   
17698        105         1    2019 2019-04-18   True           0        12   
17713        135         0    2019 2019-01-03  False           1        10   
17722        104         1    2016 2016-02-10   True           0        23   

       day_code  target  
18            4       0  
27            2       1  
77            0       1  
112           3       0  
118           5       0  
...         ...     ...  
17688         4       0  
17694         6       0  
17698         3       1  
17713         3       0  
17722         2       1  

[655 rows x 152 columns]

The games of GSW (Golden State Warriors) if the prediction is on the fourth game the model should look at the previous three games to help make better predictions. For example if the team is on a losing streak or a winning streak.

In [37]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    #closed = 'left' only have the rolling averages of the previous weeks not the week we are trying to calculate for
    rolling_stats = group[cols].rolling(3,closed='left').mean()
    group[new_cols] = rolling_stats 
    #removes all of the rows that have missing values 
    group = group.dropna(subset=new_cols)
    return group

In [38]:
cols = ["fg","3p","ft","ast","stl","blk","tov", "pts"]
new_cols = [f"{c}_rolling" for c in cols ]

In [39]:
new_cols

['fg_rolling',
 '3p_rolling',
 'ft_rolling',
 'ast_rolling',
 'stl_rolling',
 'blk_rolling',
 'tov_rolling',
 'pts_rolling']

This is just for the the GSW games 

In [40]:
rolling_averages(group, cols, new_cols)

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  day_code  \
1561   240  43   84  0.512  11   25  0.440  22   30  0.733  ...         0   
8061   240  39   85  0.459  10   26  0.385  24   31  0.774  ...         2   
13705  240  46   93  0.495  17   37  0.459  10   18  0.556  ...         4   
7734   240  38   92  0.413   8   39  0.205  19   22  0.864  ...         5   
77     240  43   91  0.473  12   21  0.571  11   13  0.846  ...         0   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
9063   240  39   86  0.453  15   37  0.405  14   20  0.700  ...         6   
16380  240  36   78  0.462  15   40  0.375  13   15  0.867  ...         2   
972    240  40   91  0.440  15   43  0.349  12   15  0.800  ...         4   
11533  240  41   88  0.466   9   40  0.225  13   15  0.867  ...         0   
7476   240  38   92  0.413  19   46  0.413   8    8  1.000  ...         3   

       target  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
1561        1   43.333333   11.666667   20.666667    27.000000     9.000000   
8061        1   44.000000   12.333333   21.333333    28.000000     9.000000   
13705       1   42.666667   12.666667   23.666667    28.666667     8.000000   
7734        1   42.666667   12.666667   18.666667    33.333333     8.000000   
77          1   41.000000   11.666667   17.666667    31.666667     9.666667   
...       ...         ...         ...         ...          ...          ...   
9063        1   41.666667   14.333333   14.666667    28.666667     6.333333   
16380       0   41.000000   16.000000   13.666667    28.333333     9.666667   
972         1   38.000000   16.333333   12.666667    23.666667    10.000000   
11533       1   38.333333   15.000000   13.000000    22.333333    10.666667   
7476        1   39.000000   13.000000   12.666667    21.666667     8.666667   

       blk_rolling  tov_rolling  pts_rolling  
1561      5.000000    12.666667   119.000000  
8061      7.000000    10.666667   121.666667  
13705     7.333333    12.000000   121.666667  
7734      8.000000    15.333333   116.666667  
77        4.333333    16.666667   111.333333  
...            ...          ...          ...  
9063      4.333333    10.666667   112.333333  
16380     3.666667    12.000000   111.666667  
972       4.333333    14.000000   105.000000  
11533     4.000000    14.666667   104.666667  
7476      4.000000    12.666667   103.666667  

[652 rows x 160 columns]

After getting the original nba_games dataframe and grouped it by teams which creates one data frame for each team in the data. Then the lambda function was applied to compute the rolling averages.

In [41]:
games_rolling = games.groupby("team").apply(lambda x:rolling_averages(x, cols, new_cols)) 

Rolling averages is applied to the entire data set 

In [42]:
games_rolling 

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  \
team                                                             ...   
ATL  16278  240  37   88  0.420   7   29  0.241  13   14  0.929  ...   
     15680  240  37   90  0.411   7   21  0.333  17   22  0.773  ...   
     9779   240  37   76  0.487   8   22  0.364  19   24  0.792  ...   
     2860   240  41   77  0.532   9   21  0.429  30   35  0.857  ...   
     4159   240  43   92  0.467  13   33  0.394  15   15  1.000  ...   
...         ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...   
WAS  12460  240  38   82  0.463   8   21  0.381  18   20  0.900  ...   
     15420  240  56  101  0.554  13   30  0.433   7   14  0.500  ...   
     9110   240  42   86  0.488  10   35  0.286   9   11  0.818  ...   
     14751  240  35   69  0.507   9   28  0.321  13   17  0.765  ...   
     1132   240  42   94  0.447   8   23  0.348  16   26  0.615  ...   

            day_code  target  fg_rolling  3p_rolling  ft_rolling  ast_rolling  \
team                                                                            
ATL  16278         6       1   38.333333    8.666667   15.666667    23.666667   
     15680         1       1   38.333333    8.333333   16.000000    23.666667   
     9779          2       1   36.666667    7.333333   15.666667    23.333333   
     2860          4       1   37.000000    7.333333   16.333333    24.666667   
     4159          5       1   38.333333    8.000000   22.000000    27.333333   
...              ...     ...         ...         ...         ...          ...   
WAS  12460         6       0   44.666667   12.000000   17.333333    31.000000   
     15420         1       1   44.666667   12.666667   19.333333    29.666667   
     9110          2       0   47.666667   12.666667   15.000000    30.666667   
     14751         4       0   45.333333   10.333333   11.333333    28.333333   
     1132          6       0   44.333333   10.666667    9.666667    28.333333   

            stl_rolling  blk_rolling  tov_rolling  pts_rolling  
team                                                            
ATL  16278     9.666667     4.000000    15.000000   101.000000  
     15680     9.666667     4.666667    13.666667   101.000000  
     9779      9.333333     3.666667    13.666667    96.333333  
     2860     11.333333     5.666667    13.000000    97.666667  
     4159     11.666667     5.666667    14.333333   106.666667  
...                 ...          ...          ...          ...  
WAS  12460     6.000000     6.333333    11.333333   118.666667  
     15420     5.333333     4.666667    10.666667   121.333333  
     9110      5.000000     3.333333    10.333333   123.000000  
     14751     4.000000     3.333333    12.000000   112.333333  
     1132      4.333333     4.333333    15.333333   109.000000  

[17682 rows x 160 columns]

dropping the team name as a seperate level in the pandas index. for a multi level index you would need to use two values to access each row which was not needed

In [43]:
games_rolling = games_rolling.droplevel('team')

In [44]:
games_rolling 

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  day_code  \
16278  240  37   88  0.420   7   29  0.241  13   14  0.929  ...         6   
15680  240  37   90  0.411   7   21  0.333  17   22  0.773  ...         1   
9779   240  37   76  0.487   8   22  0.364  19   24  0.792  ...         2   
2860   240  41   77  0.532   9   21  0.429  30   35  0.857  ...         4   
4159   240  43   92  0.467  13   33  0.394  15   15  1.000  ...         5   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
12460  240  38   82  0.463   8   21  0.381  18   20  0.900  ...         6   
15420  240  56  101  0.554  13   30  0.433   7   14  0.500  ...         1   
9110   240  42   86  0.488  10   35  0.286   9   11  0.818  ...         2   
14751  240  35   69  0.507   9   28  0.321  13   17  0.765  ...         4   
1132   240  42   94  0.447   8   23  0.348  16   26  0.615  ...         6   

       target  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
16278       1   38.333333    8.666667   15.666667    23.666667     9.666667   
15680       1   38.333333    8.333333   16.000000    23.666667     9.666667   
9779        1   36.666667    7.333333   15.666667    23.333333     9.333333   
2860        1   37.000000    7.333333   16.333333    24.666667    11.333333   
4159        1   38.333333    8.000000   22.000000    27.333333    11.666667   
...       ...         ...         ...         ...          ...          ...   
12460       0   44.666667   12.000000   17.333333    31.000000     6.000000   
15420       1   44.666667   12.666667   19.333333    29.666667     5.333333   
9110        0   47.666667   12.666667   15.000000    30.666667     5.000000   
14751       0   45.333333   10.333333   11.333333    28.333333     4.000000   
1132        0   44.333333   10.666667    9.666667    28.333333     4.333333   

       blk_rolling  tov_rolling  pts_rolling  
16278     4.000000    15.000000   101.000000  
15680     4.666667    13.666667   101.000000  
9779      3.666667    13.666667    96.333333  
2860      5.666667    13.000000    97.666667  
4159      5.666667    14.333333   106.666667  
...            ...          ...          ...  
12460     6.333333    11.333333   118.666667  
15420     4.666667    10.666667   121.333333  
9110      3.333333    10.333333   123.000000  
14751     3.333333    12.000000   112.333333  
1132      4.333333    15.333333   109.000000  

[17682 rows x 160 columns]

In [45]:
games_rolling.index = range(games_rolling.shape[0])

In [46]:
games_rolling

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  day_code  \
0      240  37   88  0.420   7   29  0.241  13   14  0.929  ...         6   
1      240  37   90  0.411   7   21  0.333  17   22  0.773  ...         1   
2      240  37   76  0.487   8   22  0.364  19   24  0.792  ...         2   
3      240  41   77  0.532   9   21  0.429  30   35  0.857  ...         4   
4      240  43   92  0.467  13   33  0.394  15   15  1.000  ...         5   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17677  240  38   82  0.463   8   21  0.381  18   20  0.900  ...         6   
17678  240  56  101  0.554  13   30  0.433   7   14  0.500  ...         1   
17679  240  42   86  0.488  10   35  0.286   9   11  0.818  ...         2   
17680  240  35   69  0.507   9   28  0.321  13   17  0.765  ...         4   
17681  240  42   94  0.447   8   23  0.348  16   26  0.615  ...         6   

       target  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
0           1   38.333333    8.666667   15.666667    23.666667     9.666667   
1           1   38.333333    8.333333   16.000000    23.666667     9.666667   
2           1   36.666667    7.333333   15.666667    23.333333     9.333333   
3           1   37.000000    7.333333   16.333333    24.666667    11.333333   
4           1   38.333333    8.000000   22.000000    27.333333    11.666667   
...       ...         ...         ...         ...          ...          ...   
17677       0   44.666667   12.000000   17.333333    31.000000     6.000000   
17678       1   44.666667   12.666667   19.333333    29.666667     5.333333   
17679       0   47.666667   12.666667   15.000000    30.666667     5.000000   
17680       0   45.333333   10.333333   11.333333    28.333333     4.000000   
17681       0   44.333333   10.666667    9.666667    28.333333     4.333333   

       blk_rolling  tov_rolling  pts_rolling  
0         4.000000    15.000000   101.000000  
1         4.666667    13.666667   101.000000  
2         3.666667    13.666667    96.333333  
3         5.666667    13.000000    97.666667  
4         5.666667    14.333333   106.666667  
...            ...          ...          ...  
17677     6.333333    11.333333   118.666667  
17678     4.666667    10.666667   121.333333  
17679     3.333333    10.333333   123.000000  
17680     3.333333    12.000000   112.333333  
17681     4.333333    15.333333   109.000000  

[17682 rows x 160 columns]

In [47]:
def make_predictions(data,predictors):
    train = data[data["date"]<'2022-01-01']
    test = data[data["date"]>'2022-01-01']
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)
    precision = precision_score(test["target"], preds)
    return combined, precision 

In [48]:
combined, precision = make_predictions(games_rolling, predictors + new_cols)

In [49]:
precision

0.5450236966824644

the precision has improved by 2%

In [50]:
combined

actual  predicted
533         0          0
534         1          0
535         0          0
536         0          1
537         0          0
...       ...        ...
17677       0          0
17678       1          0
17679       0          0
17680       0          1
17681       0          0

[1580 rows x 2 columns]

In [51]:
#merges rows based on index
combined = combined.merge(games_rolling[["date", "team", "team_opp", "won"]], left_index=True, right_index=True)

In [52]:
#information about the team that played their opponent and the result
combined

actual  predicted       date team team_opp    won
533         0          0 2022-01-03  ATL      POR  False
534         1          0 2022-01-05  ATL      SAC   True
535         0          0 2022-01-07  ATL      LAL  False
536         0          1 2022-01-09  ATL      LAC  False
537         0          0 2022-01-12  ATL      MIA  False
...       ...        ...        ...  ...      ...    ...
17677       0          0 2022-04-03  WAS      BOS  False
17678       1          0 2022-04-05  WAS      MIN   True
17679       0          0 2022-04-06  WAS      ATL  False
17680       0          1 2022-04-08  WAS      NYK  False
17681       0          0 2022-04-10  WAS      CHO  False

[1580 rows x 6 columns]

Combining home and away predictions

In [53]:
#to make sure the names of the home and away team are the same 
#the Dataset already has the same names for home and away teams
#for future reference if names need to be normalised
class MissingDict(dict):
    __missing__ = lambda self, key: key
    
map_values = {
    
    "Golden State Warriors": "GSW",
    "Lakers": "LAL",
    "Boston Celtics": "BOS"
    
}
mapping = MissingDict(**map_values)

In [54]:
mapping["Golden State Warriors"]

'GSW'

In [55]:
combined["new_team"] = combined["team"].map(mapping)

In [56]:
combined

actual  predicted       date team team_opp    won new_team
533         0          0 2022-01-03  ATL      POR  False      ATL
534         1          0 2022-01-05  ATL      SAC   True      ATL
535         0          0 2022-01-07  ATL      LAL  False      ATL
536         0          1 2022-01-09  ATL      LAC  False      ATL
537         0          0 2022-01-12  ATL      MIA  False      ATL
...       ...        ...        ...  ...      ...    ...      ...
17677       0          0 2022-04-03  WAS      BOS  False      WAS
17678       1          0 2022-04-05  WAS      MIN   True      WAS
17679       0          0 2022-04-06  WAS      ATL  False      WAS
17680       0          1 2022-04-08  WAS      NYK  False      WAS
17681       0          0 2022-04-10  WAS      CHO  False      WAS

[1580 rows x 7 columns]

This makes sure that the predictions are the same on both sides 

In [57]:

merged = combined.merge(combined, left_on=["date","new_team"], right_on=["date","team_opp"])

In [58]:
merged

actual_x  predicted_x       date team_x team_opp_x  won_x new_team_x  \
0            0            0 2022-01-03    ATL        POR  False        ATL   
1            1            0 2022-01-05    ATL        SAC   True        ATL   
2            0            0 2022-01-07    ATL        LAL  False        ATL   
3            0            1 2022-01-09    ATL        LAC  False        ATL   
4            0            0 2022-01-12    ATL        MIA  False        ATL   
...        ...          ...        ...    ...        ...    ...        ...   
1575         0            0 2022-04-03    WAS        BOS  False        WAS   
1576         1            0 2022-04-05    WAS        MIN   True        WAS   
1577         0            0 2022-04-06    WAS        ATL  False        WAS   
1578         0            1 2022-04-08    WAS        NYK  False        WAS   
1579         0            0 2022-04-10    WAS        CHO  False        WAS   

      actual_y  predicted_y team_y team_opp_y  won_y new_team_y  
0            1            1    POR        ATL   True        POR  
1            0            0    SAC        ATL  False        SAC  
2            1            1    LAL        ATL   True        LAL  
3            1            1    LAC        ATL   True        LAC  
4            1            1    MIA        ATL   True        MIA  
...        ...          ...    ...        ...    ...        ...  
1575         1            1    BOS        WAS   True        BOS  
1576         0            0    MIN        WAS  False        MIN  
1577         1            1    ATL        WAS   True        ATL  
1578         1            0    NYK        WAS   True        NYK  
1579         1            0    CHO        WAS   True        CHO  

[1580 rows x 13 columns]

In [59]:
merged[(merged["predicted_x"] == 1) & (merged["predicted_y"] == 0)]["actual_x"].value_counts()

1    291
0    215
Name: actual_x, dtype: int64

In [60]:
291/506

0.575098814229249

The accuracy is 57.5% from merging both sides of the match together 
Using more seasons could improve accuracy
Using more columns in the data frame can improve accuracy 
Look at both current team and opponent team records for the season

In [61]:
#from sklearn.metrics import f1_score

In [1]:
#f1 = f1_score(y_train, y_test, average='weighted')